src/main/java/org/jitsi/utils/dsi/DominantSpeakerIdentification.java

```diff
@@ -431,6 +482,96 @@ private synchronized Speaker getOrCreateSpeaker(T id)
         return speaker;
     }
 
+    private synchronized void updateLoudestList(Speaker speaker, int level, long now)
+    {
+        long oldestValid = now - energyExpireTimeMs;
+
+        /* exponential smoothing */
+        speaker.energyScore = (energyAlphaPct * level + (100 - energyAlphaPct) * speaker.energyScore) / 100;
+
+        logger.debug("Want to add " + speaker.id.toString() + " with score " + speaker.energyScore + ". Last level = " + level + ".");
+
+        int i = 0;
+        while (i < loudest.size())
+        {
+            Speaker cur = loudest.get(i);
+            if (cur.getLastLevelChangedTime() < oldestValid)
+            {
+                logger.debug("Removing " + cur.id.toString() + ". old.");
+                loudest.remove(i);
+                continue;
+            }
+            if (cur == speaker)
+            {
+                logger.debug("Removing " + cur.id.toString() + ". same.");
+                loudest.remove(i);
+                continue;
+            }
+            ++i;
+        }
+
+        i = 0;
+        while (i < loudest.size())
+        {
+            Speaker cur = loudest.get(i);
+            if (cur.energyScore < speaker.energyScore)
+                break;
+            ++i;
+        }
+
+        if (i < numLoudestToRoute)
+        {
+            logger.debug("Adding " + speaker.id.toString() + " at position " + i + ".");
+            loudest.add(i, speaker);
+        }
+
+        while (loudest.size() > numLoudestToRoute)
+            loudest.remove(numLoudestToRoute);
+
+        i = 0;
+        while (i < loudest.size())
+        {
+            Speaker cur = loudest.get(i);
+            logger.debug("New list: " + i + ": " + cur.id.toString() + ": " + cur.energyScore + ".");
+            ++i;
+        }
+    }
+
+    public synchronized boolean isAmongLoudestSpeakers(T id)
+    {
+        if (alwaysRouteDominant && dominantId != null && dominantId.equals(id))
+        {
+            logger.debug("Found " + id + ". Dominant speaker.");
+            return true;
+        }
+
+        int i = 0;
+        while (i < loudest.size())
+        {
+            Speaker cur = loudest.get(i);
+            if (cur.id.equals(id))
+            {
+                logger.debug("Found " + id + " at location " + i + ".");
+                return true;
+            }
+            ++i;
+        }
+
+        logger.debug(id + " not found.");
+
+        BucketStats tossedPacketsEnergyStats = this.tossedPacketsEnergyStats;
```

Comment:
This feels awkward. Why is the `isAmongLoudestSpeakers` function responsible for updating the dropped packet stats? I guess the reason is that you need access to `energyScore`, but is this really what you want to keep track of (as opposed to simply the level of the discaded packet)? It would be cleaner if the code dropping the packet updated the stats.

Reply:
now returning a class with ranking info. should be a cleaner interface and better if there are future additions.